File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        res = -float('inf')
        def dfs(node): 
            nonlocal res
            # return zero if node is null
            if not node :
                return  0
            # get the max left and right subTree while ignoring negatives
            left = max(dfs(node.left) , 0)
            right = max(dfs(node.right) , 0)
            # what is the max now ? prev or extending this path
            res = max(res , node.val + left + right)
            # one side will coninue 
            return node.val + max(left, right)
        dfs(root)
        return res

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = INT_MIN;
    int dfs(TreeNode* root){
        if (!root)
            return 0 ;
        // left path or right path
        int left = max(dfs(root->left) , 0);
        int right = max(dfs(root->right) , 0);
        // current res or extend path
        res = max(res , root->val + right + left);
        // choose one side
        return root->val + max(right ,left);
    
    }
    int maxPathSum(TreeNode* root) {
        dfs(root);
        return res;
        
    }
};